### ISSUE TITLE
Seq++ crashing on load if looping and run are enabled

### ISSUE THREAD

**Issue by nickfeisst** [squinkylabs/SquinkyVCV#197]
Over the last months I've managed to narrow down this bug that I believe is in the Seq++ module. I believe that it can crash either:

1) On load, if both looping and run are enabled when the patch is saved.
2) After launch, if Seq++ has been loaded with looping enabled, and run is then enabled.

I've attached sample patches, using only built-in modules and Seq++, to show this. The one with looping enabled crashes for me when I start Seq++ running.

[Seq++ example files.zip](https://github.com/squinkylabs/SquinkyVCV/files/6097823/Seq%2B%2B.example.files.zip)

In the past I've had to manually edit my .vcv files to get them to load, removing all connections to the Seq++ module. It's only this morning that I thought that it might be linked to the looping mode (which I often use while figuring out notes for some cover versions of songs that I've been making).

I'm not sure the details of my system will matter, but I'm using Rack v1.1.6 with macOS 11.0.1 and a 2020 M1 MacBook Pro.

**Comment by squinkylabs** [comment]
Oh, interesting. would it be possible to upload the vcv log file after a crash?

**Comment by squinkylabs** [comment]
I tried loading your patches on rack 1.16 for windows, they both loaded fine, although the loop range looks a little odd. I'lll see if I can borrow my wife's mac to duplicate this issue, but it would be really great to get a crash log from you.

**Comment by nickfeisst** [comment]
Sorry, just to be clear: the example files that I sent to you will load fine for me. However, if you push the trigger (that I've connected to the Run input of Seq++) in the file that has been saved with looping enabled, it crashes for me.

**Comment by nickfeisst** [comment]
I've just tried this on my 2019 MacBook Pro, and I think that it's not actually crashing, but hanging instead.

I've attached a 3rd .vcv file that will cause both of my computers to hang after loading the file and attempting to add/remove a cable. This file has been saved while the sequencer has both the loop and run states enabled, and will only hang after it is loaded.

The Rack log file and Apple report that are included in this archive are after opening this 3rd patch in Rack and then attempting to remove a cable. It doesn't appear to matter whether or not that cable is connected to Seq++.

[hang.zip](https://github.com/squinkylabs/SquinkyVCV/files/6099248/hang.zip)

**Comment by squinkylabs** [comment]
ah, yes, that happens for me, too, on windows. Will investigate. thanks.

**Comment by nickfeisst** [comment]
Great - thank you!

**Comment by squinkylabs** [comment]
This is in interesting challenge. thank you.

**Comment by squinkylabs** [comment]
Ok, this was fun. I found several bugs all related to looping:
1) loop range is not drawn correctly on the screen (recent regression in last release).
2) loop range has never been saved correctly to patch file.
3) If a loop range has zero time in it, Sequencer playback hangs.

This last one (#3) is what is causing your issue, but the reason the loop range has zero time in it is because of bug #2. and bug #1 was pretty obvious from looking at the screen.

I fixed all three of these, but won't me making an official release until some other things are done. Would you like me to post a mac build for you? Or do you build your own plugins?

Thanks again for tracking this down.

**Comment by nickfeisst** [comment]
That's great to hear - thank you!

I've built the plug-in and tested this just now. I can confirm that it now opens the 3rd file (that would cause the hang) and I can move cables and play the sequence without Rack hanging. It's also showing the correct loop range (I always wondered what that number was supposed to represent).

**Comment by squinkylabs** [comment]
so I assume you built it from the current "development" branch, called "mar"? Must be, as that's where the fix lives. You will probably have one or two unfinished modules in that branch, btw.

Yeah, long story on that crazy number in the loop, but won't happen again:
```c++
static void test7()
{
    SqStream s;
    s.add('x');
    std::string a = s.str();
    assertEQ(a, "x");
}
```

**Comment by squinkylabs** [comment]
This was fixed in release 1.0.16, although that version introduced a crash that will be gone in 1.0.17
